activation/nipost.go

```diff
@@ -372,28 +375,45 @@ func (nb *NIPostBuilder) submitPoetChallenge(ctx context.Context, client PoetPro
 }
 
 // Submit the challenge to all registered PoETs.
-func (nb *NIPostBuilder) submitPoetChallenges(ctx context.Context, prefix, challenge []byte, signature types.EdSignature, nodeID types.NodeID) []types.PoetRequest {
+func (nb *NIPostBuilder) submitPoetChallenges(ctx context.Context, deadline time.Time, prefix, challenge []byte, signature types.EdSignature, nodeID types.NodeID) ([]types.PoetRequest, error) {
 	g, ctx := errgroup.WithContext(ctx)
-	poetRequestsChannel := make(chan types.PoetRequest, len(nb.poetProvers))
+	type submitResult struct {
+		request *types.PoetRequest
+		err     error
+	}
+	poetRequestsChannel := make(chan submitResult, len(nb.poetProvers))
 	for _, poetProver := range nb.poetProvers {
 		poet := poetProver
 		g.Go(func() error {
-			if poetRequest, err := nb.submitPoetChallenge(ctx, poet, prefix, challenge, signature, nodeID); err == nil {
-				poetRequestsChannel <- *poetRequest
-			} else {
-				nb.log.With().Warning("failed to submit challenge to PoET", log.Err(err))
+			poetRequest, err := nb.submitPoetChallenge(ctx, deadline, poet, prefix, challenge, signature, nodeID)
+			poetRequestsChannel <- submitResult{
+				request: poetRequest,
+				err:     err,
 			}
 			return nil
 		})
 	}
 	g.Wait()
 	close(poetRequestsChannel)
 
+	allTooLate := true
 	poetRequests := make([]types.PoetRequest, 0, len(nb.poetProvers))
-	for request := range poetRequestsChannel {
-		poetRequests = append(poetRequests, request)
+	for result := range poetRequestsChannel {
+		if result.err == nil {
+			poetRequests = append(poetRequests, *result.request)
+			allTooLate = false
+		} else {
+			nb.log.With().Warning("failed to submit challenge to poet", log.Err(result.err))
+			if !errors.Is(result.err, ErrSubmitTooLate) {
+				allTooLate = false
+			}
+		}
```

Comment:
NIT: less nesting is usually easier to read 🙂 

```suggestion
		if result.err == nil {
			poetRequests = append(poetRequests, *result.request)
			allTooLate = false
			continue
		}
		
		nb.log.With().Warning("failed to submit challenge to poet", log.Err(result.err))
		if !errors.Is(result.err, ErrSubmitTooLate) {
			allTooLate = false
		}
```